### PR TITLE
refactor: convert clippy allows to clippy expects

### LINKED ISSUE #3969: 📎 Replace Clippy `allow` with `expect`
### Description

Rust 1.81 stabilizes Clippy's  [`expect` attribute](https://blog.rust-lang.org/2024/09/05/Rust-1.81.0.html#expectlint). This attribute ensures that a lint diagnostic is emitted where the attribute appears.

We should replace `allow` attribute with `expect`.

We should also retsrict the use of `allow` by denying [clippy::allow_attributes](https://rust-lang.github.io/rust-clippy/master/index.html#/allow_attributes).

### PR DESCRIPTION
**Pull request by janie314**
## Summary (Purpose)

Clippy's `allow` method allows the next line of code to be exempted from a given lint (e.g. Clippy's [unnecessary_mut_passed](https://rust-lang.github.io/rust-clippy/master/#unnecessary_mut_passed) lint).

Clippy's `expect` method is similar, except the next line of code is required for fail the given lint. This:

- Prevents unnecessary use of `allow` (`expect`), so as many lines of code are subjected to as much linting as possible.
- Ensures that `allow` (`expect`) usages are removed as soon as they're no longer needed. 

Closes #3969.

## Test Plan

Ensure CI passes, confirm code diffs are reasonable.

## Summary (Code Changes)

- Convert all uses of `allow` to `expect`.
- Remove unnecessary `expect`s.
- Adjust code generation and CI to eliminate failed `expect`s in generated code.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ unused_lifetimes                       = "warn"
 unused_macro_rules                     = "warn"
 
 [workspace.lints.clippy]
+allow_attributes        = "deny"
```
Comment (dyc3) [inline Cargo.toml:16]: I'm not entirely sure if we should outright block allow. We might want to allow `allow` in generated code to minimize new surprise diagnostics from clippy when we upgrade rust versions. Could we perhaps do that on a module level for generated code and block it everywhere else?

Comment (janie314) [reply Cargo.toml:16]: @dyc3 - I don't know-- my changes are finally finalized with `expect` globally disabled. The diff is +288, -2400, mostly removing unneeded `allow`s. My intuition says that most of the pain of switching to `expect` is contained in this repo, and that `expect` will work fine for future Clippy checks we wish to disable.

Comment (janie314) [reply Cargo.toml:16]: And it's ready for review now, @dyc3! Thank you for taking a look.

File: .github/workflows/pull_request.yml
Comment (dyc3) [inline .github/workflows/pull_request.yml:1]: The change to this workflow seems unrelated, is it necessary?

File: crates/biome_cli/tests/commands/format.rs [hunk 2]
```diff
@@ -180,14 +180,10 @@ const APPLY_ATTRIBUTE_POSITION_AFTER: &str = r#"<Foo
...
 
-// Without this, Test (windows-latest) fails with: `warning: constant `DEFAULT_CONFIGURATION_BEFORE` is never used`
-#[allow(dead_code)]
 const DEFAULT_CONFIGURATION_BEFORE: &str = r#"function f() {
     return { a, b }
   }"#;
 
-// Without this, Test (windows-latest) fails with: `warning: constant `DEFAULT_CONFIGURATION_AFTER` is never used`
-#[allow(dead_code)]
 const DEFAULT_CONFIGURATION_AFTER: &str = "function f() {
```
Comment (dyc3) [inline crates/biome_cli/tests/commands/format.rs:187]: Looks like we lost these comments - are they still relevant?

File: xtask/codegen/src/generate_syntax_kinds.rs [hunk 3]
```diff
@@ -200,8 +200,8 @@ pub fn generate_syntax_kinds(grammar: KindsSrc, language_kind: LanguageKind) ->
     };
 
     let ast = quote! {
-        #![allow(clippy::all)]
-        #![allow(bad_style, missing_docs, unreachable_pub)]
+        #![expect(clippy::all)]
+        #![expect(bad_style)]
```
Comment (dyc3) [inline xtask/codegen/src/generate_syntax_kinds.rs:204]: This is the kind of thing that I would actually prefer keeping our existing `allow`s instead. Since the `expect` is inside generated code, these diagnostics will bubble up create extra friction when setting up new languages (which is already quite high, and we're still planning to add more languages). Not ideal for DX.

File: xtask/libs_bench/benches/contains_criterion.rs [hunk 4]
```diff
@@ -4,12 +4,23 @@ use contains::*;
...
+    "undefined",
+    "a",
+    "NaN",
+    "longVariableName",
+    "Infinity",
+    "xxxxxxxx",
+    "arguments",
+    "eval",
+];
+
```
Comment (dyc3) [inline xtask/libs_bench/benches/contains_criterion.rs:17]: This was probably a function originally to prevent the compiler from optimizing the loop in these benchmarks. Could be wrong though, I'd like someone else to weigh in.

File: crates/biome_graphql_formatter/src/prelude.rs [hunk 5]
```diff
@@ -1,13 +1,9 @@
...
-#[allow(unused_imports)]
 pub(crate) use crate::{
     AsFormat, FormatNodeRule, FormattedIterExt as _, GraphqlFormatContext, GraphqlFormatter,
     IntoFormat,
 };
 pub(crate) use biome_formatter::prelude::*;
-#[allow(unused_imports)]
-pub(crate) use biome_rowan::{
-    AstNode as _, AstNodeList as _, AstNodeSlotMap as _, AstSeparatedList as _,
-};
```
Comment (dyc3) [inline crates/biome_graphql_formatter/src/prelude.rs:13]: The `allow`s here in these preludes could make sense to keep for better DX when working on formatters.

**Review by dyc3 — CHANGES_REQUESTED** [review]
This one's a doozy. Thanks for the hard work!

File: crates/biome_cli/tests/commands/format.rs (hunk 2 above)
Comment (janie314) [reply crates/biome_cli/tests/commands/format.rs:187]: The Windows test still passed with both lines removed, so it seemed not necessary.

File: .github/workflows/pull_request.yml
Comment (janie314) [reply .github/workflows/pull_request.yml:1]: Tell me what you think-- I added a `cargo fix` command to the `just gen-all` task [here](https://github.com/biomejs/biome/pull/4791/files#diff-deb9bb56fb122db0b605aa5b63f95a4665c905b18dd670e1fa6c877576a94ff1R28). This makes auto-fixes to the generated code which ensures the generated code (no longer protected by blanked `allow`s), has a clean lint.

Overall, the parts of my PR that are not simple find/replace or removing unecessary `mut`s, are:

1. This change-- adding `cargo fix --allow-dirty` to `just gen-all`.
1. Tweaking some code generation to avoid unnecessary imports and similar lint failures, [like this part](https://github.com/biomejs/biome/pull/4791/files#diff-288f5b3a2420199143fabf95327fca80b522f1f7f4c1d6d87d93242595a4cd8eR888). But sounds like you'd rather we just do some more blanket allows for those cases.

Overall, let me know what you think-- I'll be taking a look back later today if you're online today!

**Comment by janie314** [comment]
Update: I'm making several updates to this in response to @dyc3 's feedback!

File: .github/workflows/pull_request.yml
Comment (ematipico) [reply .github/workflows/pull_request.yml:1]: If possible, I'd prefer to **not** have workflows that are able to commit code in PRs, for various reasons:
- it requires more permissions in the PRs, and it might not work on PRs from external contributors (CI is hard)
- it could litter PRs with automated commits. Even if we squash the commits, people might not want those commits in the history of their PRs
- generally, contributors should be free to commit whatever they want, how they want

Comment (janie314) [reply .github/workflows/pull_request.yml:1]: I removed this-- changed my approach based no other feedback. (For what it's worth, it did not auto-commit.)

File: Cargo.toml (hunk 1 above)
Comment (janie314) [reply Cargo.toml:16]: @dyc3 I left this here. I think it works out OK because this lint does not prevent `#![allow(...)]` in individual code files. Let me know if you disagree though. I reverted many of my tweaks to the code generation process, although the process did help me implement a few lints on generated code still.

**Review by ematipico — APPROVED** [review]

File: crates/biome_js_analyze/src/lint/correctness/use_import_extensions.rs [hunk 6]
```diff
@@ -106,7 +106,7 @@ declare_lint_rule! {
     /// ## Caveats
     ///
     /// If you are using TypeScript, TypeScript version 5.0 and later is required, also make sure to enable
-    /// [allowImportingTsExtensions=true](https://typescriptlang.org/tsconfig#allowImportingTsExtensions) in your `tsconfig.json`.
+    /// [expectImportingTsExtensions=true](https://typescriptlang.org/tsconfig#allowImportingTsExtensions) in your `tsconfig.json`.
```
Comment (dyc3) [inline crates/biome_js_analyze/src/lint/correctness/use_import_extensions.rs:109]: This looks like a find replace accident
```suggestion
    /// [allowImportingTsExtensions=true](https://typescriptlang.org/tsconfig#allowImportingTsExtensions) in your `tsconfig.json`.
```

**Review by dyc3 — CHANGES_REQUESTED** [review]
Looks good, just a minor nit

**Review by dyc3 — APPROVED** [review]
